build/dart/rules.gni

```diff
@@ -5,7 +5,178 @@
 # This file has rules for making Dart packages and Dart-based Mojo applications.
 # The entrypoint is the dart_pkg rule.
 
+import("//build/compiled_action.gni")
 import("//build/module_args/dart.gni")
+import("//flutter/common/config.gni")
+import("//third_party/dart/build/dart/dart_action.gni")
+
+# Creates a dart kernel (dill) file suitalbe for use with gen_snapshot.
+#
+# Invoker must supply dart_main and package_config. Invoker may optionally
+# supply aot as a boolean and product as a boolean.
+template("dart_snapshot") {
+  assert(defined(invoker.main_dart), "main_dart is a required parameter.")
+  assert(defined(invoker.sources), "sources is a required parameter.")
+  assert(defined(invoker.package_config),
+         "package_config is a required parameter.")
+
+  kernel_target = "_${target_name}_kernel"
+  snapshot_target = "_${target_name}_snapshot"
+  is_aot =
+      flutter_runtime_mode == "profile" || flutter_runtime_mode == "release"
+
+  kernel_output = "$target_gen_dir/kernel_blob.bin"
+
+  prebuilt_dart_action(kernel_target) {
+    main_dart = rebase_path(invoker.main_dart)
+    package_config = rebase_path(invoker.package_config)
+
+    platform_dill = "$root_out_dir/flutter_patched_sdk/platform_strong.dill"
+    gen_kernel_script = "//third_party/dart/pkg/vm/bin/gen_kernel.dart"
+
+    deps = [ "//flutter/lib/snapshot:strong_platform" ]
+
+    inputs = [
+               main_dart,
+               package_config,
+               platform_dill,
+               gen_kernel_script,
+             ] + invoker.sources
+
+    outputs = [ kernel_output ]
+
+    depfile = "$kernel_output.d"
+    abs_depfile = rebase_path(depfile)
+    rebased_output = rebase_path(kernel_output, root_build_dir)
+    vm_args = [
+      "--depfile=$abs_depfile",
+      "--depfile_output_filename=$rebased_output",
+      "--disable-dart-dev",
+    ]
+
+    script = gen_kernel_script
+
+    args = [
+      "--packages=" + rebase_path(package_config),
+      "--target=flutter",
+      "--platform=" + rebase_path(platform_dill),
+      "--output=" + rebase_path(kernel_output),
+    ]
+
+    if (is_debug) {
```

Comment:
I am confused if you meant `is_debug` to mean `unopt` instead of the debug runtime mode.

Reply:
We should probably rename this to `is_unopt`. But that's what it means - it's distinct from the runtime mode.